### PR TITLE
Add preview of components to documentation website

### PR DESCRIPTION
**Pull request by frankieroberto**
This lets you see what each component looks like, as well as the code to generate it, in a similar way to the [GOV.UK Design System website](https://design-system.service.gov.uk) and [GOV.UK Component for Ruby on Rails website](https://govuk-components.netlify.app).

#### Before

Screenshot 2023-12-01 at 12 50 59

#### After

Screenshot 2023-12-01 at 12 50 30

### CODE REVIEW

File: docs/examples/masthead.njk [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+title: Masthead example
+layout: example.njk
+eleventyExcludeFromCollections: true
+
```
Comment (paulrobertlloyd) [inline docs/examples/masthead.njk:5]:
```suggestion
```

File: docs/examples/autocomplete.njk [hunk 2]
```diff
@@ -0,0 +1,46 @@
+---
+title: Autocomplete example
+layout: example.njk
+eleventyExcludeFromCollections: true
+---
+{% from "x-govuk/components/autocomplete/macro.njk" import xGovukAutocomplete %}
```
Comment (paulrobertlloyd) [inline docs/examples/autocomplete.njk:6]: Can we not import these components in the `example.njk` layout… oh remembering issue with Eleventy and macros… is this still an issue?

File: docs/index.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+Two JavaScript-only modules are also provided:
+
+| Module | Description |
+| - | - |
+| [Edge](/edge) | Define the edges of your prototype for research. |
+| [Warn on unsaved changes](/warn-on-unsaved-changes) | Warn users if they try to leave a page without saving changes to a form. |
+
+> [!NOTE]\
+> Prior to v1.0.0, this project included a collection of decorated form components. These can now be found in the [`govuk-decorated-components`](https://github.com/x-govuk/govuk-decorated-components) package.
```
Comment (paulrobertlloyd) [inline docs/index.md:29]: Given the markdown renderer we’re using doesn’t understand this syntax, and that we’re now using this separate from the main README, we can use this instead:

```suggestion
> **NOTE**
> Prior to v1.0.0, this project included a collection of decorated form components. These can now be found in the [`govuk-decorated-components`](https://github.com/x-govuk/govuk-decorated-components) package.
```

File: docs/index.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
+---
+homepage: true
+layout: product
+includeInBreadcrumbs: true
+title: GOV.UK Prototype Components
```
Comment (paulrobertlloyd) [inline docs/index.md:5]: We’d probably want to follow the examples of other X-GOVUK sites and move some of the content on this page to a ‘Get started’ page, but we can do that in a subsequent PR.

File: docs/package.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (paulrobertlloyd) [inline docs/package.json:1]: Do we need a separate `package.json` for the docs? On the other projects we add dev dependencies and scripts for building/previewing the docs. Probably simpler.

File: docs/index.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+| - | - |
+| [Edge](/edge) | Define the edges of your prototype for research. |
+| [Warn on unsaved changes](/warn-on-unsaved-changes) | Warn users if they try to leave a page without saving changes to a form. |
+
+> [!NOTE]\
+> Prior to v1.0.0, this project included a collection of decorated form components. These can now be found in the [`govuk-decorated-components`](https://github.com/x-govuk/govuk-decorated-components) package.
+
+## Requirements
+
+Node.js v16 or later.
```
Comment (paulrobertlloyd) [inline docs/index.md:33]: I think this needs updating, as currently the package.json for this project has `"node": ">=18.0.0"` (which also needs to be updated to account for v20). Perhaps one for a separate PR?

File: docs/index.md (hunk 4 above)
Comment (paulrobertlloyd) [reply docs/index.md:5]: Also, perhaps after his is merge, we can update the main `README` to instead link to such a page, save having content be duplicated in 2 places.

File: docs/eleventy.config.js [hunk 7]
```diff
@@ -31,33 +34,35 @@ module.exports = function (eleventyConfig) {
...
+    let data = matter(fs.readFileSync(`examples/${componentName}.njk`, 'utf-8')).content
+
+    data = data.replace(/\{\%\sfrom\s[^\n]+\n\n/, '')
 
-  // Transforms
-  eleventyConfig.addTransform('remove-h1', (content, outputPath) => {
-    // Remove first `h1` as it repeats what’s already shown in page title
-    content = content.replace(/<h1\s*.*tabindex="-1"\s*.*>\s*.*<\/h1>/, '')
-    return content
+    const highlightedCode = hljs.highlight(data, { language: 'js' }).value
```
Comment (paulrobertlloyd) [inline docs/eleventy.config.js:49]: Where are we using this? We already include [the highlighter in `markdown-it-govuk`](https://github.com/x-govuk/markdown-it-govuk#code-highlighting) which is depended on by the Eleventy plugin, so maybe we can use that instead?

**Review by paulrobertlloyd — CHANGES_REQUESTED** [review]
This is God’s work @frankieroberto! Thanks for taking it on. Left a few initial comments, will wait until these and the other todo items are done before taking a closer look.

P.S. Guess we’re gunna need a homepage illustration, huh? 😊

**Comment by paulrobertlloyd** [comment]
As to resizing `iframe`s, I think the GOV.UK docs use https://github.com/davidjbradshaw/iframe-resizer

File: docs/examples/autocomplete.njk (hunk 2 above)
Comment (frankieroberto) [reply docs/examples/autocomplete.njk:6]: Tried that. Didn’t work and couldn’t work out why. Any ideas?

File: docs/package.json (hunk 5 above)
Comment (frankieroberto) [reply docs/package.json:1]: I was assuming we’d have a separate `package.json` and update the build script. I think they need to be separate as a package can’t depend on itself?

File: docs/index.md (hunk 4 above)
Comment (frankieroberto) [reply docs/index.md:5]: Ah yes, I forgot to remove the duplicate content from the README. I've changed that now in this branch to just have a link to the docs.

And yeah, possibly that content could get split up into a "Get started" page, with separate pages for contributing and releasing (unless we move that stuff back to the README).

File: docs/index.md (hunk 6 above)
Comment (frankieroberto) [reply docs/index.md:33]: Opps, well spotted. Fixed the initial mis-match in #179.

Not sure if we should also update to minimum 20 or not? Are we dependent on a version of the prototype kit which depends on v20?

File: docs/eleventy.config.js [hunk 8]
```diff
@@ -31,33 +34,35 @@ module.exports = function (eleventyConfig) {
...
   })
 
-  // Ignores
-  eleventyConfig.ignores.add('**/*/*.js')
-  eleventyConfig.ignores.add('**/*/*.scss')
-  eleventyConfig.ignores.add('**/*/*.njk')
+  eleventyConfig.addNunjucksGlobal("getNunjucksCode", function(componentName) {
+    let data = matter(fs.readFileSync(`examples/${componentName}.njk`, 'utf-8')).content
+
+    data = data.replace(/\{\%\sfrom\s[^\n]+\n\n/, '')
```
[the remark this answers is not in the document]
Comment (frankieroberto) [reply docs/eleventy.config.js]: I've added a comment to try and explain this in 33596b17317e43191863bbea2d6d034a223ed7b8:

> This function fetches the raw nunjucks code for the given example and returns it.
> It’s needed as the `{% include "" %}` feature within nunjucks will render the
> code, and there’s currently no built-in way to fetch it un-rendered.
>
> See https://github.com/mozilla/nunjucks/issues/788

I mainly lifted it from the Design System website code: [exports.getNunjucksCode](https://github.com/alphagov/govuk-design-system/blob/main/lib/nunjucks/globals.js#L14-L25)

**Comment by frankieroberto** [comment]
@paulrobertlloyd thanks for the first look. I've addressed most of the comments, but still need to figure out why the autocomplete isn't working (any ideas?) and how to get `iframe-resizer` working (can't immediately see how to add javascript to the `govuk-eleventy-plugin`).

**Review by paulrobertlloyd — APPROVED** [review]
Done with my fiddling, and got it working with one `package.json` and so similar set up to its sister project repos.

Feel free to squash and merge!

**Pull request merged**